### ISSUE TITLE
Feature request: Allow general callbacks for validation

### ISSUE THREAD

**Issue by fuglede** [chriskiehl/Gooey#293]
Currently, a `GooeyParser` can take a general string for validation, and this string will then be `eval`'ed when necessary. While this works for simple stuff, anything that requires more than Python built-ins becomes hard to express. For example, it could be natural to test `os.path.isfile` on the contents of a `FileChooser`, but that's not currently possible (since you would have to `import os` at some point).

Instead, it would be nice to be able to provide a completely general Python callable for use in validation. One possible API would be to simply allow the existing `TextContainer._options['validator']['test']` to be a callable and check its type when necessary. Otherwise, one monkey patch workaround is given in the following example in which we extend `TextContainer._options['validator']` to allow callables as well (but in the process break the ordinary validation):

```python
from gooey.gui.components.widgets.bases import TextContainer
oldGetValue = TextContainer.getValue

def newGetValue(self):
    result = oldGetValue(self)
    userValidator = self._options['validator']['callback']
    message = self._options['validator']['message']
    value = self.getWidgetValue()
    validates = userValidator(value)
    result['test'] = validates
    result['error'] = None if validates else message
    return result

TextContainer.getValue = newGetValue

@Gooey
def main():
    parser = GooeyParser()
    parser.add_argument('stuff', widget='FileChooser', gooey_options={
        'validator':{
            'callback': os.path.isfile,
            'message': 'some helpful message'
        }
    })
    args = parser.parse_args()
```

**Comment by bilderbuchi** [comment]
You could probably have both ways by supplying the eval approach with a lambda as a default for a keyword argument that can be overridden with a user callable?

**Comment by fuglede** [comment]
@bilderbuchi: So if no callable is provided in what I call `'callback'` in the example, it falls back to the contents of `'test'`, otherwise ignores it? If so, makes sense. Otherwise, what I was thinking was that if you wanted to keep it as a single option `'test'` (since you'd never use them at the same time anyway), you could, in [`TextContainer.getValue`](https://github.com/chriskiehl/Gooey/blob/b19fe8826dfae85c013506b4d49bbaee478e2006/gooey/gui/components/widgets/bases.py#L99), just do something to the extent of
```python
testFunc = userValidator if callable(userValidator) else eval(f'lambda user_input: bool({userValidator})')`
```

**Comment by chriskiehl** [comment]
Ok, so, this is something that's actually not currently possible with Gooey, but would make an absolutely _awesome_ feature. However, implementing it will be non-trivial as it would involve a fair bit of hijacking of the client code, _plus_ probably some re-thinking of some of Gooey's main build spec. 

**On why you can't just supply a callable:**

Gooey doesn't call any of your code directly nor even hold a reference to it. Gooey is run entirely independently from a plain of json file. The only thing it knows about your code is its location on disk so that it can call it with `Popen`. 

The main things that the 'bindings' do (e.g. the decorator and all the `gooey_options` stuff) is just generate that json spec file and feed it to Gooey. So, if it cannot be serialized (like arbitrary python code), it cannot be stored in that JSON file. Pickle is an obvious work around to this problem, but that would bypass the spirit of why Gooey reads from a language agnostic data format to begin with: I want Gooey to (eventually) work with any language. Making Gooey read from a plain ol' data file was step one towards this goal. 

**On Hijacking:**

To get the arbitrary callable thing to work, we'd have to do a setup similar to the current [Dynamic values](https://github.com/chriskiehl/gooey#using-dynamic-values) one. Meaning, Gooey would have to operate kind of like a client/server thing. The Hijacking part is that Gooey would need to put a new validation command line argument _in the client code_. Similar to how the [Dynamic values](https://github.com/chriskiehl/gooey#using-dynamic-values)  or the `--ignore-gooey` flag works now. If Gooey requests validation, it would go through a code path that maps `input -> your validation function`, give back some structured response, and then BAM feature achieved.

**Comment by fuglede** [comment]
@chriskiehl: Thanks for the reply. I'm new to Gooey and so I can't claim to understand how its bindings work (at all), but I'm not sure I follow: In the example above we do actually supply a callable, and since it is being invoked at the level of the bindings themselves, it just works. On the other hand, regarding having it work with any language, I suppose it is also problematic that the expression currently given to `'test'` is a Python one?

**Comment by Magnati** [comment]
@chriskiehl 
I agree with @fuglede that I do not see the problem to enhance the code here:

I changed only getValue in the following way. And for me it works.
Am I missing something? If not, I sent a pull request.

```
    def getValue(self):
        userValidatorCB = None
        validatesCB = True
        value = self.getWidgetValue()
        if 'callback' in self._options['validator']:
            userValidatorCB = self._options['validator']['callback']
            validatesCB = userValidatorCB(value)

        userValidator = getin(self._options, ['validator', 'test'], 'True')
        message = getin(self._options, ['validator', 'message'], '')
        testFunc = eval('lambda user_input: bool(%s)' % userValidator)
        satisfies = testFunc if self._meta['required'] else ifPresent(testFunc)
        result = {
            'id': self._id,
            'cmd': self.formatOutput(self._meta, value),
            'rawValue': value,
            'test': runValidator(satisfies, value),
            'error': None if runValidator(satisfies, value) and validatesCB else message,
            'clitype': 'positional'
                        if self._meta['required'] and not self._meta['commands']
                        else 'optional'
        }
        return result
```

**Comment by chriskiehl** [comment]
To understand why it doesn't work, you need only enable to `load_build_config` config option to force Gooey to read from a serialized json file rather than bootstrap itself from memory. There is no generalized serializable format for a Python callable, so you can't arbitrarily store them. 

So, in short, as I mentioned above, this feature would break Gooey's ability to boostrap itself from a plain ol' language agnostic json file. 

Just to clear up the rationale a bit: 

Outside of my work on Gooey, I don't write that much Python anymore (I've been lured away by glorious Clojure ^_^) By drawing a firm line in the sand between Gooey and the program it runs, I'm able to still use Gooey even though I'm in a totally different programming ecosystem/environment. Making this interop easier is the direction Gooey is heading.

Passing a callable directly would be _easy_ for Python programmers in the short term, but would limit Gooey's long term usefulness. 

All that said, the feature is still 100% possible, and the API can be exactly what you want: passing a callable to the parser. It's just that in order to make that work, and still have Gooey remain language agnostic, it'd have to inject a custom root level parser argument and associate that function with it so that it can be called in a client/server setup ala [Dynamic Values](https://github.com/chriskiehl/gooey#using-dynamic-values)

I'm likely explaining the steps required to make it work very poorly. However, they're floating in my head and on my list of things to do!

**Comment by Magnati** [comment]
Well, okay then. 
Looking forward for the enhanced version.

**Comment by blipk** [comment]
Would it be possible to implement an easier fix to pass import options to gooey_options validator so os.path.isfile etc can be used?
